clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -1,17 +1,210 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 
 public class VeniceAfterImageConsumerImpl<K, V> extends VeniceChangelogConsumerImpl<K, V> {
+  private static final Logger LOGGER = LogManager.getLogger(VeniceAfterImageConsumerImpl.class);
+  // 10 Minute default
+  protected long versionSwapDetectionIntervalTimeInMs;
+  // This consumer is used to find EOP messages without impacting consumption by other subscriptions. It's only used
+  // in the context of seeking to EOP in the event of the user calling that seek or a version push.
+  // TODO: We shouldn't use this in the long run. Once the EOP position is queryable from venice and version
+  // swap is produced to VT, then we should remove this as it's no longer needed.
+  final private Lazy<VeniceChangelogConsumerImpl<K, V>> internalSeekConsumer;
+  private Thread versionSwapDetectionThread;
+
   public VeniceAfterImageConsumerImpl(ChangelogClientConfig changelogClientConfig, PubSubConsumerAdapter consumer) {
+    this(
+        changelogClientConfig,
+        consumer,
+        Lazy.of(
+            () -> new VeniceChangelogConsumerImpl<K, V>(
+                changelogClientConfig,
+                VeniceChangelogConsumerClientFactory.getConsumer(
+                    changelogClientConfig.getConsumerProperties(),
+                    changelogClientConfig.getStoreName() + "-" + "internal"))));
+  }
+
+  protected VeniceAfterImageConsumerImpl(
+      ChangelogClientConfig changelogClientConfig,
+      PubSubConsumerAdapter consumer,
+      Lazy<VeniceChangelogConsumerImpl<K, V>> seekConsumer) {
     super(changelogClientConfig, consumer);
+    versionSwapDetectionThread = new VersionSwapDetectionThread();
+    internalSeekConsumer = seekConsumer;
+    versionSwapDetectionIntervalTimeInMs = changelogClientConfig.getVersionSwapDetectionIntervalTimeInMs();
   }
 
   @Override
   public Collection<PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate>> poll(long timeoutInMs) {
+    if (!versionSwapDetectionThread.isAlive()) {
+      versionSwapDetectionThread.start();
+    }
     return internalPoll(timeoutInMs, "");
   }
+
+  @Override
+  public CompletableFuture<Void> seekToTimestamps(Map<Integer, Long> timestamps) {
+    return internalSeekToTimestamps(timestamps, "");
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToTail(Set<Integer> partitions) {
+    return internalSeekToTail(partitions, "");
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToEndOfPush(Set<Integer> partitions) {
+    return CompletableFuture.supplyAsync(() -> {
+      synchronized (internalSeekConsumer) {
+        try {
+          // TODO: This implementation basically just scans the version topic until it finds the EOP message. The
+          // approach
+          // we'd like to do is instead add the offset of the EOP message in the VT, and then just seek to that offset.
```

Comment:
Not related to this change, wondering how would you add the offset of EOP in the VT so that we can easily to query?

Reply:
I'm going to add it into a system store and make it queryable for the client.